```
Error running mint automation
+ mkdir -p 16222-d2fbed1/gopath/src/github.com/minio
+ git -C 16222-d2fbed1/gopath/src/github.com/minio clone --quiet https://github.com/minio/minio.git
+ git -C 16222-d2fbed1/gopath/src/github.com/minio/minio remote add minio https://github.com/minio/minio.git
+ git -C 16222-d2fbed1/gopath/src/github.com/minio/minio fetch --quiet minio pull/16222/head:pr16222
+ git -C 16222-d2fbed1/gopath/src/github.com/minio/minio checkout --quiet pr16222
+ GO111MODULE=on
+ GOPATH=/home/mint-auto/mint-auto/16222-d2fbed1/gopath
+ make -C 16222-d2fbed1/gopath/src/github.com/minio/minio --quiet
go: downloading cloud.google.com/go/storage v1.26.0
go: downloading github.com/minio/madmin-go/v2 v2.0.1
go: downloading github.com/klauspost/readahead v1.4.0
go: downloading github.com/minio/minio-go/v7 v7.0.44
go: downloading github.com/bits-and-blooms/bloom/v3 v3.3.1
go: downloading github.com/cespare/xxhash/v2 v2.1.2
go: downloading github.com/klauspost/reedsolomon v1.11.3
go: downloading github.com/Azure/azure-storage-blob-go v0.15.0
go: downloading github.com/cosnicolaou/pbzip2 v1.0.1
go: downloading github.com/dchest/siphash v1.2.3
go: downloading github.com/cheggaaa/pb v1.0.29
go: downloading github.com/djherbis/atime v1.1.0
go: downloading github.com/bcicen/jstream v1.0.1
go: downloading github.com/coreos/go-oidc v2.2.1+incompatible
go: downloading github.com/minio/dperf v0.4.2
go: downloading github.com/fatih/color v1.13.0
go: downloading github.com/minio/cli v1.24.0
go: downloading github.com/klauspost/pgzip v1.2.5
go: downloading github.com/dustin/go-humanize v1.0.0
go: downloading github.com/minio/console v0.22.2
go: downloading github.com/google/uuid v1.3.0
go: downloading github.com/felixge/fgprof v0.9.3
go: downloading github.com/klauspost/filepathx v1.1.1
go: downloading github.com/klauspost/compress v1.15.12
go: downloading github.com/gorilla/mux v1.8.0
go: downloading github.com/miekg/dns v1.1.50
go: downloading github.com/json-iterator/go v1.1.12
go: downloading github.com/minio/pkg v1.5.8
go: downloading github.com/minio/highwayhash v1.0.2
go: downloading github.com/minio/kes v0.22.0
go: downloading github.com/go-openapi/loads v0.21.2
go: downloading github.com/lithammer/shortuuid/v4 v4.0.0
go: downloading github.com/hashicorp/golang-lru v0.5.4
go: downloading github.com/golang-jwt/jwt/v4 v4.4.2
go: downloading github.com/mitchellh/go-homedir v1.1.0
go: downloading github.com/minio/sio v0.3.0
go: downloading github.com/prometheus/client_golang v1.13.0
go: downloading github.com/minio/selfupdate v0.5.0
go: downloading github.com/inconshreveable/mousetrap v1.0.1
go: downloading github.com/minio/xxml v0.0.3
go: downloading github.com/minio/zipindex v0.3.0
go: downloading github.com/philhofer/fwd v1.1.2-0.20210722190033-5c56ac6d0bb9
go: downloading github.com/prometheus/client_model v0.3.0
go: downloading github.com/pierrec/lz4 v2.6.1+incompatible
go: downloading golang.org/x/crypto v0.3.0
go: downloading github.com/prometheus/procfs v0.8.0
go: downloading github.com/rs/cors v1.8.2
go: downloading github.com/rs/dnscache v0.0.0-20211102005908-e0241e321417
go: downloading go.etcd.io/etcd/client/v3 v3.5.5
go: downloading github.com/zeebo/xxh3 v1.0.2
go: downloading go.etcd.io/etcd/api/v3 v3.5.5
go: downloading github.com/secure-io/sio-go v0.3.1
go: downloading go.uber.org/atomic v1.10.0
go: downloading github.com/valyala/bytebufferpool v1.0.0
go: downloading github.com/shirou/gopsutil/v3 v3.22.9
go: downloading github.com/prometheus/common v0.37.0
go: downloading github.com/tinylib/msgp v1.1.7-0.20220719154719-f3635b96e483
go: downloading golang.org/x/oauth2 v0.2.0
go: downloading golang.org/x/sys v0.2.0
go: downloading google.golang.org/api v0.98.0
go: downloading gopkg.in/yaml.v2 v2.4.0
go: downloading github.com/buger/jsonparser v1.1.1
go: downloading github.com/go-ldap/ldap/v3 v3.4.4
go: downloading github.com/coredns/coredns v1.9.4
go: downloading golang.org/x/time v0.0.0-20220722155302-e5dcc9cfc0b9
go: downloading go.uber.org/zap v1.23.0
go: downloading github.com/beevik/ntp v0.3.0
go: downloading github.com/ncw/directio v1.0.5
go: downloading github.com/minio/sha256-simd v1.0.0
go: downloading github.com/Shopify/sarama v1.36.0
go: downloading github.com/eclipse/paho.mqtt.golang v1.4.1
go: downloading github.com/elastic/go-elasticsearch/v7 v7.17.7
go: downloading github.com/go-sql-driver/mysql v1.6.0
go: downloading github.com/gomodule/redigo v1.8.9
go: downloading github.com/lib/pq v1.10.7
go: downloading github.com/minio/simdjson-go v0.4.2
go: downloading github.com/nats-io/nats.go v1.17.0
go: downloading github.com/nats-io/stan.go v0.10.3
go: downloading github.com/nsqio/go-nsq v1.1.0
go: downloading github.com/pkg/errors v0.9.1
go: downloading github.com/rabbitmq/amqp091-go v1.5.0
go: downloading github.com/xdg/scram v1.0.5
go: downloading github.com/bits-and-blooms/bitset v1.3.3
go: downloading github.com/golang/protobuf v1.5.2
go: downloading golang.org/x/sync v0.1.0
go: downloading github.com/mattn/go-colorable v0.1.13
go: downloading github.com/mattn/go-isatty v0.0.16
go: downloading github.com/mattn/go-runewidth v0.0.14
go: downloading github.com/pquerna/cachecontrol v0.1.0
go: downloading gopkg.in/square/go-jose.v2 v2.6.0
go: downloading github.com/gogo/protobuf v1.3.2
go: downloading github.com/fraugster/parquet-go v0.12.0
go: downloading github.com/minio/csvparser v1.0.0
go: downloading github.com/alecthomas/participle v0.7.1
go: downloading golang.org/x/net v0.2.0
go: downloading github.com/google/pprof v0.0.0-20220829040838-70bd9ae97f40
go: downloading github.com/klauspost/cpuid/v2 v2.2.1
go: downloading github.com/Azure/go-ntlmssp v0.0.0-20220621081337-cb9428e4ac1e
go: downloading github.com/go-asn1-ber/asn1-ber v1.5.4
go: downloading github.com/golang/snappy v0.0.4
go: downloading github.com/nats-io/nuid v1.0.1
go: downloading go.etcd.io/etcd/client/pkg/v3 v3.5.5
go: downloading google.golang.org/grpc v1.50.1
go: downloading github.com/go-openapi/analysis v0.21.4
go: downloading github.com/go-openapi/spec v0.20.7
go: downloading github.com/go-openapi/swag v0.22.3
go: downloading github.com/xdg/stringprep v1.0.3
go: downloading github.com/rivo/uniseg v0.4.2
go: downloading github.com/coreos/go-semver v0.3.0
go: downloading google.golang.org/genproto v0.0.0-20221018160656-63c7b68cfc55
go: downloading github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
go: downloading github.com/modern-go/reflect2 v1.0.2
go: downloading github.com/matttproud/golang_protobuf_extensions v1.0.2
go: downloading github.com/Azure/azure-pipeline-go v0.2.3
go: downloading github.com/gorilla/websocket v1.5.0
go: downloading github.com/beorn7/perks v1.0.1
go: downloading google.golang.org/protobuf v1.28.1
go: downloading github.com/montanaflynn/stats v0.6.6
go: downloading github.com/fatih/structs v1.1.0
go: downloading github.com/lestrrat-go/jwx v1.2.25
go: downloading github.com/rjeczalik/notify v0.9.2
go: downloading github.com/rs/xid v1.4.0
go: downloading gopkg.in/ini.v1 v1.67.0
go: downloading github.com/minio/md5-simd v1.1.2
go: downloading aead.dev/mem v0.2.0
go: downloading github.com/coreos/go-systemd/v22 v22.4.0
go: downloading github.com/nats-io/nkeys v0.3.0
go: downloading github.com/tklauser/go-sysconf v0.3.11
go: downloading golang.org/x/text v0.4.0
go: downloading cloud.google.com/go v0.104.0
go: downloading github.com/mattn/go-ieproxy v0.0.1
go: downloading go.uber.org/multierr v1.8.0
go: downloading github.com/mailru/easyjson v0.7.7
go: downloading gopkg.in/yaml.v3 v3.0.1
go: downloading github.com/apache/thrift v0.16.0
go: downloading github.com/rcrowley/go-metrics v0.0.0-20201227073835-cf1acfcdf475
go: downloading github.com/davecgh/go-spew v1.1.1
go: downloading github.com/tklauser/numcpus v0.6.0
go: downloading github.com/jcmturner/gokrb5/v8 v8.4.3
go: downloading github.com/jcmturner/gofork v1.7.6
go: downloading github.com/eapache/queue v1.1.0
go: downloading github.com/eapache/go-resiliency v1.3.0
go: downloading github.com/hashicorp/go-multierror v1.1.1
go: downloading github.com/pierrec/lz4/v4 v4.1.16
go: downloading github.com/eapache/go-xerial-snappy v0.0.0-20180814174437-776d5712da21
go: downloading github.com/josharian/intern v1.0.0
go: downloading github.com/hashicorp/errwrap v1.1.0
go: downloading github.com/go-openapi/strfmt v0.21.3
go: downloading github.com/go-openapi/jsonpointer v0.19.5
go: downloading github.com/lestrrat-go/backoff/v2 v2.0.8
go: downloading github.com/lestrrat-go/option v1.0.0
go: downloading github.com/lestrrat-go/blackmagic v1.0.1
go: downloading github.com/lestrrat-go/iter v1.0.2
go: downloading github.com/lestrrat-go/httpcc v1.0.1
go: downloading github.com/go-openapi/jsonreference v0.20.0
go: downloading github.com/jcmturner/dnsutils/v2 v2.0.0
go: downloading github.com/hashicorp/go-uuid v1.0.3
go: downloading github.com/jcmturner/aescts/v2 v2.0.0
go: downloading github.com/jcmturner/rpc/v2 v2.0.3
go: downloading github.com/asaskevich/govalidator v0.0.0-20210307081110-f21760c49a8d
go: downloading github.com/oklog/ulid v1.3.1
go: downloading github.com/go-openapi/errors v0.20.3
go: downloading go.mongodb.org/mongo-driver v1.10.3
go: downloading github.com/mitchellh/mapstructure v1.5.0
go: downloading cloud.google.com/go/iam v0.4.0
go: downloading go.opencensus.io v0.23.0
go: downloading golang.org/x/xerrors v0.0.0-20220907171357-04be3eba64a2
go: downloading github.com/googleapis/gax-go/v2 v2.5.1
go: downloading cloud.google.com/go/compute v1.10.0
go: downloading github.com/golang/groupcache v0.0.0-20210331224755-41bb18bfe9da
go: downloading github.com/go-openapi/runtime v0.24.2
go: downloading github.com/jessevdk/go-flags v1.5.0
go: downloading github.com/tidwall/gjson v1.14.3
go: downloading github.com/go-openapi/validate v0.22.0
go: downloading github.com/minio/mc v0.0.0-20221201184114-854b4f123f03
go: downloading github.com/unrolled/secure v1.13.0
go: downloading github.com/tidwall/pretty v1.2.1
go: downloading github.com/tidwall/match v1.1.1
go: downloading github.com/docker/go-units v0.5.0
go: downloading github.com/jedib0t/go-pretty/v6 v6.3.8
go: downloading github.com/charmbracelet/bubbles v0.14.0
go: downloading gopkg.in/h2non/filetype.v1 v1.0.5
go: downloading github.com/gdamore/tcell/v2 v2.5.3
go: downloading github.com/charmbracelet/lipgloss v0.6.0
go: downloading github.com/google/shlex v0.0.0-20191202100458-e7afc7fbc510
go: downloading github.com/charmbracelet/bubbletea v0.22.1
go: downloading github.com/rivo/tview v0.0.0-20220916081518-2e69b7385a37
go: downloading github.com/minio/colorjson v1.0.4
go: downloading github.com/prometheus/prom2json v1.3.2
go: downloading github.com/pkg/xattr v0.4.9
go: downloading github.com/minio/filepath v1.0.0
go: downloading github.com/navidys/tvxwidgets v0.1.1
go: downloading golang.org/x/term v0.2.0
go: downloading github.com/muesli/reflow v0.3.0
go: downloading github.com/posener/complete v1.2.3
go: downloading github.com/olekukonko/tablewriter v0.0.5
go: downloading github.com/muesli/termenv v0.13.0
go: downloading github.com/lucasb-eyer/go-colorful v1.2.0
go: downloading github.com/aymanbagabas/go-osc52 v1.2.1
go: downloading github.com/mattn/go-localereader v0.0.1
go: downloading github.com/containerd/console v1.0.3
go: downloading github.com/muesli/cancelreader v0.2.2
go: downloading github.com/muesli/ansi v0.0.0-20211031195517-c9f0611b6c70
go: downloading github.com/gdamore/encoding v1.0.0
go: downloading github.com/googleapis/enterprise-certificate-proxy v0.1.0
go: downloading github.com/google/go-cmp v0.5.9
# github.com/minio/minio/internal/http
internal/http/transports.go:61:6: tlsClientConfig declared but not used
make: *** [build] Error 2

```

--- a/cmd/common-main.go
+++ b/cmd/common-main.go
@@ -134,7 +134,7 @@ func init() {
 
 	globalForwarder = handlers.NewForwarder(&handlers.Forwarder{
 		PassHost:     true,
-		RoundTripper: newHTTPTransport(1 * time.Hour),
+		RoundTripper: NewHTTPTransportWithTimeout(1 * time.Hour),
 		Logger: func(err error) {
 			if err != nil && !errors.Is(err, context.Canceled) {
 				logger.LogIf(GlobalContext, err)

--- a/cmd/config-current.go
+++ b/cmd/config-current.go
@@ -551,7 +551,7 @@ func applyDynamicConfigForSubSys(ctx context.Context, objAPI ObjectLayer, s conf
 
 		// Initialize remote instance transport once.
 		getRemoteInstanceTransportOnce.Do(func() {
-			getRemoteInstanceTransport = newHTTPTransport(apiConfig.RemoteTransportDeadline)
+			getRemoteInstanceTransport = NewHTTPTransportWithTimeout(apiConfig.RemoteTransportDeadline)
 		})
 	case config.CompressionSubSys:
 		cmpCfg, err := compress.LookupConfig(s[config.CompressionSubSys][config.Default])

--- a/cmd/server-main.go
+++ b/cmd/server-main.go
@@ -19,7 +19,6 @@ package cmd
 
 import (
 	"context"
-	"crypto/tls"
 	"errors"
 	"fmt"
 	"io"
@@ -42,10 +41,8 @@ import (
 	"github.com/minio/minio/internal/bucket/bandwidth"
 	"github.com/minio/minio/internal/color"
 	"github.com/minio/minio/internal/config"
-	"github.com/minio/minio/internal/fips"
 	xhttp "github.com/minio/minio/internal/http"
 	"github.com/minio/minio/internal/logger"
-	"github.com/minio/minio/internal/rest"
 	"github.com/minio/pkg/certs"
 	"github.com/minio/pkg/env"
 )
@@ -236,19 +233,9 @@ func serverHandleCmdArgs(ctx *cli.Context) {
 	}
 
 	// allow transport to be HTTP/1.1 for proxying.
-	globalProxyTransport = newCustomHTTPProxyTransport(&tls.Config{
-		RootCAs:            globalRootCAs,
-		CipherSuites:       fips.TLSCiphers(),
-		CurvePreferences:   fips.TLSCurveIDs(),
-		ClientSessionCache: tls.NewLRUClientSessionCache(tlsClientSessionCacheSize),
-	}, rest.DefaultTimeout)()
+	globalProxyTransport = NewCustomHTTPProxyTransport()()
 	globalProxyEndpoints = GetProxyEndpoints(globalEndpoints)
-	globalInternodeTransport = newInternodeHTTPTransport(&tls.Config{
-		RootCAs:            globalRootCAs,
-		CipherSuites:       fips.TLSCiphers(),
-		CurvePreferences:   fips.TLSCurveIDs(),
-		ClientSessionCache: tls.NewLRUClientSessionCache(tlsClientSessionCacheSize),
-	}, rest.DefaultTimeout)()
+	globalInternodeTransport = NewInternodeHTTPTransport()()
 	globalRemoteTargetTransport = NewRemoteTargetHTTPTransport()()
 
 	// On macOS, if a process already listens on LOCALIPADDR:PORT, net.Listen() falls back

--- a/cmd/test-utils_test.go
+++ b/cmd/test-utils_test.go
@@ -66,7 +66,6 @@ import (
 	"github.com/minio/minio/internal/crypto"
 	"github.com/minio/minio/internal/hash"
 	"github.com/minio/minio/internal/logger"
-	"github.com/minio/minio/internal/rest"
 	"github.com/minio/pkg/bucket/policy"
 )
 
@@ -112,7 +111,7 @@ func TestMain(m *testing.M) {
 	// Initialize globalConsoleSys system
 	globalConsoleSys = NewConsoleLogger(context.Background())
 
-	globalInternodeTransport = newInternodeHTTPTransport(nil, rest.DefaultTimeout)()
+	globalInternodeTransport = NewInternodeHTTPTransport()()
 
 	initHelp()
 

--- a/cmd/utils.go
+++ b/cmd/utils.go
@@ -38,7 +38,6 @@ import (
 	"runtime/trace"
 	"strings"
 	"sync"
-	"syscall"
 	"time"
 
 	"github.com/coreos/go-oidc"
@@ -60,6 +59,7 @@ import (
 	"github.com/minio/minio/internal/logger"
 	"github.com/minio/minio/internal/logger/message/audit"
 	"github.com/minio/minio/internal/mcontext"
+	"github.com/minio/minio/internal/rest"
 	"github.com/minio/pkg/certs"
 	"github.com/minio/pkg/env"
 	xnet "github.com/minio/pkg/net"
@@ -568,163 +568,73 @@ func ToS3ETag(etag string) string {
 	return etag
 }
 
-func newInternodeHTTPTransport(tlsConfig *tls.Config, dialTimeout time.Duration) func() http.RoundTripper {
-	// For more details about various values used here refer
-	// https://golang.org/pkg/net/http/#Transport documentation
-	tr := &http.Transport{
-		Proxy:                 http.ProxyFromEnvironment,
-		DialContext:           xhttp.DialContextWithDNSCache(globalDNSCache, xhttp.NewInternodeDialContext(dialTimeout)),
-		MaxIdleConnsPerHost:   1024,
-		WriteBufferSize:       32 << 10, // 32KiB moving up from 4KiB default
-		ReadBufferSize:        32 << 10, // 32KiB moving up from 4KiB default
-		IdleConnTimeout:       15 * time.Second,
-		ResponseHeaderTimeout: 15 * time.Minute, // Set conservative timeouts for MinIO internode.
-		TLSHandshakeTimeout:   15 * time.Second,
-		ExpectContinueTimeout: 15 * time.Second,
-		TLSClientConfig:       tlsConfig,
-		// Go net/http automatically unzip if content-type is
-		// gzip disable this feature, as we are always interested
-		// in raw stream.
-		DisableCompression: true,
-	}
-
-	// https://github.com/golang/go/issues/23559
-	// https://github.com/golang/go/issues/42534
-	// https://github.com/golang/go/issues/43989
-	// https://github.com/golang/go/issues/33425
-	// https://github.com/golang/go/issues/29246
-	// if tlsConfig != nil {
-	// 	trhttp2, _ := http2.ConfigureTransports(tr)
-	// 	if trhttp2 != nil {
-	// 		// ReadIdleTimeout is the timeout after which a health check using ping
-	// 		// frame will be carried out if no frame is received on the
-	// 		// connection. 5 minutes is sufficient time for any idle connection.
-	// 		trhttp2.ReadIdleTimeout = 5 * time.Minute
-	// 		// PingTimeout is the timeout after which the connection will be closed
-	// 		// if a response to Ping is not received.
-	// 		trhttp2.PingTimeout = dialTimeout
-	// 		// DisableCompression, if true, prevents the Transport from
-	// 		// requesting compression with an "Accept-Encoding: gzip"
-	// 		trhttp2.DisableCompression = true
-	// 	}
-	// }
-
-	return func() http.RoundTripper {
-		return tr
-	}
+// NewInternodeHTTPTransport returns a transport for internode MinIO
+// connections.
+func NewInternodeHTTPTransport() func() http.RoundTripper {
+	return xhttp.ConnSettings{
+		DNSCache:         globalDNSCache,
+		DialTimeout:      rest.DefaultTimeout,
+		RootCAs:          globalRootCAs,
+		CipherSuites:     fips.TLSCiphers(),
+		CurvePreferences: fips.TLSCurveIDs(),
+		EnableHTTP2:      false,
+	}.NewInternodeHTTPTransport()
 }
 
-// Used by only proxied requests, specifically only supports HTTP/1.1
-func newCustomHTTPProxyTransport(tlsConfig *tls.Config, dialTimeout time.Duration) func() *http.Transport {
-	// For more details about various values used here refer
-	// https://golang.org/pkg/net/http/#Transport documentation
-	tr := &http.Transport{
-		Proxy:                 http.ProxyFromEnvironment,
-		DialContext:           xhttp.DialContextWithDNSCache(globalDNSCache, xhttp.NewInternodeDialContext(dialTimeout)),
-		MaxIdleConnsPerHost:   1024,
-		WriteBufferSize:       32 << 10, // 32KiB moving up from 4KiB default
-		ReadBufferSize:        32 << 10, // 32KiB moving up from 4KiB default
-		IdleConnTimeout:       15 * time.Second,
-		ResponseHeaderTimeout: 30 * time.Minute, // Set larger timeouts for proxied requests.
-		TLSHandshakeTimeout:   10 * time.Second,
-		ExpectContinueTimeout: 10 * time.Second,
-		TLSClientConfig:       tlsConfig,
-		// Go net/http automatically unzip if content-type is
-		// gzip disable this feature, as we are always interested
-		// in raw stream.
-		DisableCompression: true,
-	}
-
-	return func() *http.Transport {
-		return tr
-	}
-}
-
-func newCustomHTTPTransport(tlsConfig *tls.Config, dialTimeout time.Duration) func() *http.Transport {
-	// For more details about various values used here refer
-	// https://golang.org/pkg/net/http/#Transport documentation
-	tr := &http.Transport{
-		Proxy:                 http.ProxyFromEnvironment,
-		DialContext:           xhttp.DialContextWithDNSCache(globalDNSCache, xhttp.NewInternodeDialContext(dialTimeout)),
-		MaxIdleConnsPerHost:   1024,
-		WriteBufferSize:       32 << 10, // 32KiB moving up from 4KiB default
-		ReadBufferSize:        32 << 10, // 32KiB moving up from 4KiB default
-		IdleConnTimeout:       15 * time.Second,
-		ResponseHeaderTimeout: 3 * time.Minute,
-		TLSHandshakeTimeout:   10 * time.Second,
-		ExpectContinueTimeout: 10 * time.Second,
-		TLSClientConfig:       tlsConfig,
-		// Go net/http automatically unzip if content-type is
-		// gzip disable this feature, as we are always interested
-		// in raw stream.
-		DisableCompression: true,
-	}
-
-	// https://github.com/golang/go/issues/23559
-	// https://github.com/golang/go/issues/42534
-	// https://github.com/golang/go/issues/43989
-	// https://github.com/golang/go/issues/33425
-	// https://github.com/golang/go/issues/29246
-	// if tlsConfig != nil {
-	// 	trhttp2, _ := http2.ConfigureTransports(tr)
-	// 	if trhttp2 != nil {
-	// 		// ReadIdleTimeout is the timeout after which a health check using ping
-	// 		// frame will be carried out if no frame is received on the
-	// 		// connection. 5 minutes is sufficient time for any idle connection.
-	// 		trhttp2.ReadIdleTimeout = 5 * time.Minute
-	// 		// PingTimeout is the timeout after which the connection will be closed
-	// 		// if a response to Ping is not received.
-	// 		trhttp2.PingTimeout = dialTimeout
-	// 		// DisableCompression, if true, prevents the Transport from
-	// 		// requesting compression with an "Accept-Encoding: gzip"
-	// 		trhttp2.DisableCompression = true
-	// 	}
-	// }
-
-	return func() *http.Transport {
-		return tr
-	}
+// NewCustomHTTPProxyTransport is used only for proxied requests, specifically
+// only supports HTTP/1.1
+func NewCustomHTTPProxyTransport() func() *http.Transport {
+	return xhttp.ConnSettings{
+		DNSCache:         globalDNSCache,
+		DialTimeout:      rest.DefaultTimeout,
+		RootCAs:          globalRootCAs,
+		CipherSuites:     fips.TLSCiphers(),
+		CurvePreferences: fips.TLSCurveIDs(),
+		EnableHTTP2:      false,
+	}.NewCustomHTTPProxyTransport()
 }
 
 // NewHTTPTransportWithClientCerts returns a new http configuration
 // used while communicating with the cloud backends.
 func NewHTTPTransportWithClientCerts(clientCert, clientKey string) *http.Transport {
-	transport := newHTTPTransport(1 * time.Minute)
+	s := xhttp.ConnSettings{
+		DNSCache:    globalDNSCache,
+		DialTimeout: 1 * time.Minute,
+		RootCAs:     globalRootCAs,
+		EnableHTTP2: false,
+	}
+
 	if clientCert != "" && clientKey != "" {
 		ctx, cancel := context.WithTimeout(context.Background(), time.Second)
 		defer cancel()
-		c, err := certs.NewManager(ctx, clientCert, clientKey, tls.LoadX509KeyPair)
+		transport, err := s.NewHTTPTransportWithClientCerts(ctx, clientCert, clientKey)
 		if err != nil {
 			logger.LogIf(ctx, fmt.Errorf("failed to load client key and cert, please check your endpoint configuration: %s",
 				err.Error()))
 		}
-		if c != nil {
-			c.UpdateReloadDuration(10 * time.Second)
-			c.ReloadOnSignal(syscall.SIGHUP) // allow reloads upon SIGHUP
-			transport.TLSClientConfig.GetClientCertificate = c.GetClientCertificate
-		}
+		return transport
 	}
-	return transport
+
+	return s.NewHTTPTransportWithTimeout(1 * time.Minute)
 }
 
 // NewHTTPTransport returns a new http configuration
 // used while communicating with the cloud backends.
 func NewHTTPTransport() *http.Transport {
-	return newHTTPTransport(1 * time.Minute)
+	return NewHTTPTransportWithTimeout(1 * time.Minute)
 }
 
 // Default values for dial timeout
 const defaultDialTimeout = 5 * time.Second
 
-func newHTTPTransport(timeout time.Duration) *http.Transport {
-	tr := newCustomHTTPTransport(&tls.Config{
-		RootCAs:            globalRootCAs,
-		ClientSessionCache: tls.NewLRUClientSessionCache(tlsClientSessionCacheSize),
-	}, defaultDialTimeout)()
-
-	// Customize response header timeout
-	tr.ResponseHeaderTimeout = timeout
-	return tr
+// NewHTTPTransportWithTimeout allows setting a timeout.
+func NewHTTPTransportWithTimeout(timeout time.Duration) *http.Transport {
+	return xhttp.ConnSettings{
+		DNSCache:    globalDNSCache,
+		DialTimeout: defaultDialTimeout,
+		RootCAs:     globalRootCAs,
+		EnableHTTP2: false,
+	}.NewHTTPTransportWithTimeout(timeout)
 }
 
 type dialContext func(ctx context.Context, network, addr string) (net.Conn, error)
@@ -753,29 +663,11 @@ func newCustomDialContext() dialContext {
 // NewRemoteTargetHTTPTransport returns a new http configuration
 // used while communicating with the remote replication targets.
 func NewRemoteTargetHTTPTransport() func() *http.Transport {
-	// For more details about various values used here refer
-	// https://golang.org/pkg/net/http/#Transport documentation
-	tr := &http.Transport{
-		Proxy:                 http.ProxyFromEnvironment,
-		DialContext:           newCustomDialContext(),
-		MaxIdleConnsPerHost:   1024,
-		WriteBufferSize:       32 << 10, // 32KiB moving up from 4KiB default
-		ReadBufferSize:        32 << 10, // 32KiB moving up from 4KiB default
-		IdleConnTimeout:       15 * time.Second,
-		TLSHandshakeTimeout:   5 * time.Second,
-		ExpectContinueTimeout: 5 * time.Second,
-		TLSClientConfig: &tls.Config{
-			RootCAs:            globalRootCAs,
-			ClientSessionCache: tls.NewLRUClientSessionCache(tlsClientSessionCacheSize),
-		},
-		// Go net/http automatically unzip if content-type is
-		// gzip disable this feature, as we are always interested
-		// in raw stream.
-		DisableCompression: true,
-	}
-	return func() *http.Transport {
-		return tr
-	}
+	return xhttp.ConnSettings{
+		DialContext: newCustomDialContext(),
+		RootCAs:     globalRootCAs,
+		EnableHTTP2: false,
+	}.NewCustomHTTPProxyTransport()
 }
 
 // Load the json (typically from disk file).

--- a/cmd/warm-backend-minio.go
+++ b/cmd/warm-backend-minio.go
@@ -42,7 +42,7 @@ func newWarmBackendMinIO(conf madmin.TierMinIO) (*warmBackendMinIO, error) {
 	creds := credentials.NewStaticV4(conf.AccessKey, conf.SecretKey, "")
 
 	getRemoteTierTargetInstanceTransportOnce.Do(func() {
-		getRemoteTierTargetInstanceTransport = newHTTPTransport(10 * time.Minute)
+		getRemoteTierTargetInstanceTransport = NewHTTPTransportWithTimeout(10 * time.Minute)
 	})
 	opts := &minio.Options{
 		Creds:     creds,

--- a/cmd/warm-backend-s3.go
+++ b/cmd/warm-backend-s3.go
@@ -117,7 +117,7 @@ func newWarmBackendS3(conf madmin.TierS3) (*warmBackendS3, error) {
 		creds = credentials.NewStaticV4(conf.AccessKey, conf.SecretKey, "")
 	}
 	getRemoteTierTargetInstanceTransportOnce.Do(func() {
-		getRemoteTierTargetInstanceTransport = newHTTPTransport(10 * time.Minute)
+		getRemoteTierTargetInstanceTransport = NewHTTPTransportWithTimeout(10 * time.Minute)
 	})
 	opts := &minio.Options{
 		Creds:     creds,

--- a/internal/http/transports.go
+++ b/internal/http/transports.go
@@ -1,0 +1,177 @@
+// Copyright (c) 2015-2022 MinIO, Inc.
+//
+// # This file is part of MinIO Object Storage stack
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package http
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"net"
+	"net/http"
+	"syscall"
+	"time"
+
+	"github.com/minio/pkg/certs"
+	"github.com/rs/dnscache"
+)
+
+// tlsClientSessionCacheSize is the cache size for client sessions.
+var tlsClientSessionCacheSize = 100
+
+// ConnSettings - contains connection settings.
+type ConnSettings struct {
+	// If this is non-nil, DNSCache and DialTimeout are ignored.
+	DialContext func(ctx context.Context, network, addr string) (net.Conn, error)
+
+	// Dial settings, used if DialContext is nil.
+	DNSCache    *dnscache.Resolver
+	DialTimeout time.Duration
+
+	// TLS Settings
+	RootCAs          *x509.CertPool
+	CipherSuites     []uint16
+	CurvePreferences []tls.CurveID
+
+	// HTTP2
+	EnableHTTP2 bool
+}
+
+func (s ConnSettings) getDefaultTransport() *http.Transport {
+	dialContext := s.DialContext
+	if dialContext == nil {
+		dialContext = DialContextWithDNSCache(s.DNSCache, NewInternodeDialContext(s.DialTimeout))
+	}
+
+	tlsClientConfig := tls.Config{
+		RootCAs:            s.RootCAs,
+		CipherSuites:       s.CipherSuites,
+		CurvePreferences:   s.CurvePreferences,
+		ClientSessionCache: tls.NewLRUClientSessionCache(tlsClientSessionCacheSize),
+	}
+
+	// For more details about various values used here refer
+	// https://golang.org/pkg/net/http/#Transport documentation
+	tr := &http.Transport{
+		Proxy:                 http.ProxyFromEnvironment,
+		DialContext:           dialContext,
+		MaxIdleConnsPerHost:   1024,
+		WriteBufferSize:       32 << 10, // 32KiB moving up from 4KiB default
+		ReadBufferSize:        32 << 10, // 32KiB moving up from 4KiB default
+		IdleConnTimeout:       15 * time.Second,
+		ResponseHeaderTimeout: 15 * time.Minute, // Conservative timeout is the default (for MinIO internode)
+		TLSHandshakeTimeout:   10 * time.Second,
+		ExpectContinueTimeout: 10 * time.Second,
+		TLSClientConfig:       &tlsClientConfig,
+		ForceAttemptHTTP2:     s.EnableHTTP2,
+		// Go net/http automatically unzip if content-type is
+		// gzip disable this feature, as we are always interested
+		// in raw stream.
+		DisableCompression: true,
+	}
+
+	// https://github.com/golang/go/issues/23559
+	// https://github.com/golang/go/issues/42534
+	// https://github.com/golang/go/issues/43989
+	// https://github.com/golang/go/issues/33425
+	// https://github.com/golang/go/issues/29246
+	// if tlsConfig != nil {
+	// 	trhttp2, _ := http2.ConfigureTransports(tr)
+	// 	if trhttp2 != nil {
+	// 		// ReadIdleTimeout is the timeout after which a health check using ping
+	// 		// frame will be carried out if no frame is received on the
+	// 		// connection. 5 minutes is sufficient time for any idle connection.
+	// 		trhttp2.ReadIdleTimeout = 5 * time.Minute
+	// 		// PingTimeout is the timeout after which the connection will be closed
+	// 		// if a response to Ping is not received.
+	// 		trhttp2.PingTimeout = dialTimeout
+	// 		// DisableCompression, if true, prevents the Transport from
+	// 		// requesting compression with an "Accept-Encoding: gzip"
+	// 		trhttp2.DisableCompression = true
+	// 	}
+	// }
+
+	return tr
+}
+
+// NewInternodeHTTPTransport returns transport for internode MinIO connections.
+func (s ConnSettings) NewInternodeHTTPTransport() func() http.RoundTripper {
+	tr := s.getDefaultTransport()
+
+	// Settings specific to internode requests.
+	tr.TLSHandshakeTimeout = 15 * time.Minute
+	tr.ExpectContinueTimeout = 15 * time.Minute
+
+	return func() http.RoundTripper {
+		return tr
+	}
+}
+
+// NewCustomHTTPProxyTransport is used only for proxied requests, specifically
+// only supports HTTP/1.1
+func (s ConnSettings) NewCustomHTTPProxyTransport() func() *http.Transport {
+	s.EnableHTTP2 = false
+	tr := s.getDefaultTransport()
+
+	// Settings specific to proxied requests.
+	tr.ResponseHeaderTimeout = 30 * time.Minute
+
+	return func() *http.Transport {
+		return tr
+	}
+}
+
+// NewHTTPTransportWithTimeout allows setting a timeout for response headers
+func (s ConnSettings) NewHTTPTransportWithTimeout(timeout time.Duration) *http.Transport {
+	tr := s.getDefaultTransport()
+
+	// Settings specific to this transport.
+	tr.ResponseHeaderTimeout = timeout
+	return tr
+}
+
+// NewHTTPTransportWithClientCerts returns a new http configuration used for
+// communicating with client cert authentication.
+func (s ConnSettings) NewHTTPTransportWithClientCerts(ctx context.Context, clientCert, clientKey string) (*http.Transport, error) {
+	transport := s.NewHTTPTransportWithTimeout(1 * time.Minute)
+	if clientCert != "" && clientKey != "" {
+		c, err := certs.NewManager(ctx, clientCert, clientKey, tls.LoadX509KeyPair)
+		if err != nil {
+			return nil, err
+		}
+		if c != nil {
+			c.UpdateReloadDuration(10 * time.Second)
+			c.ReloadOnSignal(syscall.SIGHUP) // allow reloads upon SIGHUP
+			transport.TLSClientConfig.GetClientCertificate = c.GetClientCertificate
+		}
+	}
+	return transport, nil
+}
+
+// NewRemoteTargetHTTPTransport returns a new http configuration
+// used while communicating with the remote replication targets.
+func (s ConnSettings) NewRemoteTargetHTTPTransport() func() *http.Transport {
+	tr := s.getDefaultTransport()
+
+	tr.TLSHandshakeTimeout = 5 * time.Second
+	tr.ExpectContinueTimeout = 5 * time.Second
+	tr.ResponseHeaderTimeout = 0
+
+	return func() *http.Transport {
+		return tr
+	}
+}

